storage/remote/client.go

```diff
@@ -120,3 +130,105 @@ func (c *Client) Store(samples model.Samples) error {
 func (c Client) Name() string {
 	return fmt.Sprintf("%d:%s", c.index, c.url)
 }
+
+// Read reads from a remote endpoint.
+func (c *Client) Read(ctx context.Context, from, through model.Time, matchers metric.LabelMatchers) (model.Matrix, error) {
+	req := &ReadRequest{
+		// TODO: Support batching multiple queries into one read request,
+		// as the protobuf interface allows for it.
+		Queries: []*Query{{
+			StartTimestampMs: int64(from),
+			EndTimestampMs:   int64(through),
+			Matchers:         labelMatchersToProto(matchers),
+		}},
+	}
+
+	data, err := proto.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal read request: %v", err)
+	}
+
+	buf := bytes.Buffer{}
+	if _, err := snappy.NewWriter(&buf).Write(data); err != nil {
+		return nil, err
+	}
+
+	// TODO: POST is weird for a query, but as per http://stackoverflow.com/questions/978061/http-get-with-request-body
+	// it's not valid to attach meaning to an HTTP body in a GET request.
+	// What is the right thing to do here?
```

Comment:
POST is the done thing.

We'll be adding it to scrapes at some point too due to url length limits when ?name[]= becomes popular